### ISSUE TITLE
Incorporate docs of middleware in expressjs org and utilities in jshttp, pillarjs orgs

### ISSUE THREAD

**Issue by crandmck** [expressjs/expressjs.com#680]
I think it's commonly understood that much of the power of Express comes through middleware.  Obviously, we can't and shouldn't try to document all the middleware out there, but I'd like to improve how we document the set of middleware modules in the [expressjs](https://github.com/expressjs) and [jshttp](https://github.com/jshttp) orgs.  Together, these two orgs provide over 50 important/basic middleware packages and they are the closest thing we have to "official" middleware, though we don't refer to them that way.  

Currently we just list (some?) of these packages in seemingly-random order and link to the GitHub repos from http://expressjs.com/en/resources/middleware.html.  Middleware in expressjs and jshttp orgs are not clearly distinguished from other third-party middleware packages.

I've been thinking for some time about the best way to address this.  Obviously, each middleware needs to have its own documentation (typically in README.md), but I believe that in addition the Express docs should also provide a unified and easy way to find and use these _de facto_ "official" middleware packages. 

So, over the weekend I tried a little experiment (not rocket science, really) where I just wrote a [bash script](https://github.com/expressjs/expressjs.com/blob/mw-module-doc/get-readmes.sh) to `curl` down the README.md files for all the middleware repos in expressjs and jshttp.  Then using a Jekyll template, I incorporated this doc into pages on the site. It works surprisingly well.  For maintenance, we'd just have to periodically run the script to get any updates to the middleware package READMEs (could also just update a package's README individually). 

The "proof of concept" is currently in https://github.com/expressjs/expressjs.com/tree/mw-module-doc. There are several improvements and refinements needed, but first I wanted to get feedback on the idea before spending any more time on it.  

At a high level, the point of this exercise is to make it easier to use the _de facto_ standard set of middleware in these two orgs, without interfering with their individual package documentation.

So @hacksparrow @dougwilson and anyone else...thoughts?

**Comment by hacksparrow** [comment]
If each of these middleware doc are accompanied with a short description of what role they play in Express, technically apt users will find them particularly useful. I am all for it.

**Comment by dougwilson** [comment]
I don't see any reason why we cannot copy the READMEs onto the main expressjs.com website, as long as someone is willing to actively work on keeping it up-to-date (probably means just periodically running that script and checking in any changes, I assume). The only real comments beyond that I have is the following:
- The branch puts them all under "middlewares", though most of those modules are not middlewares (particularly nothing in the "jshttp" organization is a middleware in any way).
- Until we can clean up the "expressjs" and "pillarjs" organizations, we may want to just make a list of modules to import, rather than everything in the org. A few of the modules are abandoned, for example.
- We probably want to indicate somewhere in the copied text what version that text is for. For example, if you are just scraping the README from the `master` branch, you may end up putting a README on expressjs.com that has information not even published to npm yet.

**Comment by crandmck** [comment]
All good points.

> If each of these middleware doc are accompanied with a short description of what role they play in Express...

Yes, that's definitely one of the things I'll add.

> ... as long as someone is willing to actively work on keeping it up-to-date (probably means just periodically running that script and checking in any changes, I assume)

Yes, exactly.  I nominate myself for that duty. I'll start watching the repos, but can you give me a sense of how often these modules' external interfaces or READMEs change?  I guess it's hard to generalize, but are we talking daily, weekly, monthly, or less?  My implicit assumption is that it would be monthly or less, which wouldn't be to hard for me to keep up with. :-)

> The branch puts them all under "middlewares", though most of those modules are not middlewares (particularly nothing in the "jshttp" organization is a middleware in any way).

Oh!  Obviously, I didn't look closely at each individual module (especially the jshttp modules).  So, it looks like the jshttp modules are utility or "helper" modules?  Is that right?  I'll find a different place to list them, not under middleware.  I'll also review the expressjs modules to see which are middleware and which aren't.

@dougwilson Should I add the [pillarjs](https://github.com/pillarjs) modules? From a cursory look, it seems like these modules include both middleware and general utility modules?  Is that right?  If so, then I can go through and sort them according to usage...

> ...we may want to just make a list of modules to import, rather than everything in the org. A few of the modules are abandoned, for example.

Yes, definitely; it would be the list in the bash script.  I started with the complete list of modules and removed somet that I could tell right away were either not relevant or abandoned (obviously I missed a few).  I'll review the whole list again, but @dougwilson I would also appreciate your input as well on any others to remove from the list.

>  ...indicate somewhere in the copied text what version that text is for. For example, if you are just scraping the README from the master branch, you may end up putting a README on expressjs.com that has information not even published to npm yet.

That is a really good point.  It would make most sense to try to get the README from the current version published to npm, if that's possible.   @dougwilson is it reasonable to assume that the "latest" tagged release in GH is published to npm?  I'll take another look at the GitHub API and see if I can figure out a way to grab that README.  Or I'm open to other suggestions, thanks!

**Comment by dougwilson** [comment]
> but can you give me a sense of how often these modules' external interfaces or READMEs change? I guess it's hard to generalize, but are we talking daily, weekly, monthly, or less? My implicit assumption is that it would be monthly or less, which wouldn't be to hard for me to keep up with. :-)

Yea, I would classify them as monthly or less. A lot of the modules are "done" and so generally don't have too many changes except an occasional new feature here or a random bug/typo fix there.

> So, it looks like the jshttp modules are utility or "helper" modules? Is that right?

Yea, basically :)

> Should I add the pillarjs modules? From a cursory look, it seems like these modules include both middleware and general utility modules? Is that right? If so, then I can go through and sort them according to usage...

They would probably be the most generally useful compared to jshttp. As far as effort to add them, I would say expend the effort on expresjs, then pillarjs, then jshttp last if desired.

> but @dougwilson I would also appreciate your input as well on any others to remove from the list.

No problem! At least from your initial branch, I would not include `express-messages`, `express-resource`, `express-expose`, `express-params`, `connect-render`, `urlrouter`, `restful-router`, `domain-middleware`, `connect-markdown`, `routification`, `vhostess`, `http-utils`, and `spdy-push`.

> is it reasonable to assume that the "latest" tagged release in GH is published to npm?

Definitely. In addition, I'm going to start on some of these module trying to keep in progress changes on a non-`master` branch and see how well that works. If we end up doing this, then it makes it a non-issue :)

**Comment by crandmck** [comment]
Thanks for your clarifications @dougwilson 
I've done a bit more work on the branch, but haven't quite been able to get it to the point where I'm ready for a PR.  I did find a way to determine the "latest tagged release" via GitHub API, but it fails for some of the repos (even though they seem to have a release that should qualify).  I'm still trying to get to the bottom of that.

I'm thinking that initially I'll grab the READMEs just for the middleware modules (via script), and just link out to the utility modules in a table with a short summary of each does.

The other thing I'd like to do is use Node instead of bash to get the READMEs from GH API.  It doesn't matter too much since it's just a command-line script that's called manually from time-to-time, but I think it's the right thing to do.  I don't know that I need to hold off the PR for that though.  Let me see how far I can get :-)

**Comment by crandmck** [comment]
So, I addressed Doug's comments, and opened PR (#687).  Since I didn't hear anything for 5 days, I went ahead and merged it.   I hope that's not too controversial.... 

Basically, I just reworked http://expressjs.com/en/resources/middleware.html to highlight the `expressjs` org middleware (in a table layout), and pulled in the READMEs (via GH API) for all the modules so someone can browse their docs on this site (with a little side navbar to make it easy to browse)

I also added http://expressjs.com/en/resources/utils.html to give a brief summary of some of the utility modules in `pillarjs`.

I want to replace the bash script that grabs the READMEs via GitHub API with a Node program, and I actually wrote the code, but I need to do a bit more work on it before it's ready...   As part of that, I'd like to add a note at the top of each README that we use stating the version and date it was grabbed.  That'll be coming along in another PR as time permits.
